userspace/libsinsp/container_info.cpp

```diff
@@ -17,49 +17,246 @@ limitations under the License.
+	m_health_probe_exe = other.m_health_probe_exe;
+	m_health_probe_args = other.m_health_probe_args;
+	m_obj = other.m_obj;
+
+	return *this;
+}
+
+void sinsp_container_info::container_health_probe::add_health_probes(const Json::Value &config_obj,
+								     std::list<container_health_probe> &probes)
+{
+	Json::Value spec;
+	bool liveness_readiness_added = false;
+
+	// When parsing the full container json for live containers, a label contains stringified json that
+	// contains the probes.
+	if (container_health_probe::get_k8s_pod_spec(config_obj, spec))
 	{
-		return NULL;
+		if(spec.isMember("livenessProbe"))
+		{
+			container_health_probe probe;
+			if(probe.parse_liveness_readiness_probe(spec["livenessProbe"], PT_LIVENESS_PROBE))
+			{
+				probes.push_back(probe);
+				liveness_readiness_added = true;
+			}
+		}
+		else if(spec.isMember("readinessProbe"))
+		{
+			container_health_probe probe;
+			if(probe.parse_liveness_readiness_probe(spec["readinessProbe"], PT_READINESS_PROBE))
+			{
+				probes.push_back(probe);
+				liveness_readiness_added = true;
+			}
+		}
 	}
 
-	return &(m_mounts[idx]);
+	// When parsing the json for container_json events,
+	// the livenessProbe/readinessProbe are directly
+	// available.
+	if (config_obj.isMember("livenessProbe"))
+	{
+		container_health_probe probe;
+		if(probe.parse_liveness_readiness_probe(config_obj["livenessProbe"], PT_LIVENESS_PROBE))
+		{
+			probes.push_back(probe);
+			liveness_readiness_added = true;
+		}
+	}
+
+	if (config_obj.isMember("readinessProbe"))
+	{
+		container_health_probe probe;
+		if(probe.parse_liveness_readiness_probe(config_obj["readinessProbe"], PT_READINESS_PROBE))
+		{
+			probes.push_back(probe);
+			liveness_readiness_added = true;
+		}
+	}
+
+	// To avoid any confusion about containers that both refer to
+	// a healthcheck and liveness/readiness probe, we only
+	// consider a healthcheck if no liveness/readiness was added.
+
+	if(!liveness_readiness_added && config_obj.isMember("Healthcheck"))
+	{
+		container_health_probe probe;
+		if(probe.parse_healthcheck(config_obj["Healthcheck"]))
+		{
+			probes.push_back(probe);
+		}
+	}
 }
 
-const sinsp_container_info::container_mount_info *sinsp_container_info::mount_by_source(std::string &source) const
+void sinsp_container_info::container_health_probe::add_to_json(Json::Value &container_obj) const
 {
-	// note: linear search
-	for (auto &mntinfo :m_mounts)
+	switch(m_probe_type)
 	{
-		if(sinsp_utils::glob_match(source.c_str(), mntinfo.m_source.c_str()))
+	case PT_HEALTHCHECK:
+		container_obj["Healthcheck"] = m_obj;
+		break;
+	case PT_LIVENESS_PROBE:
+		container_obj["livenessProbe"] = m_obj;
+		break;
+	case PT_READINESS_PROBE:
+		container_obj["readinessProbe"] = m_obj;
+		break;
+	default:
+		g_logger.format(sinsp_logger::SEV_ERROR, "Unknown health probe type %d",
+				m_probe_type);
+	}
+}
+
+bool sinsp_container_info::container_health_probe::get_k8s_pod_spec(const Json::Value &config_obj, Json::Value &spec)
```

Comment:
I don't feel k8s/Docker specific code belongs here. If we already use e.g. Docker JSON as a serialized representation e.g. in scap files then we're stuck with it, but otherwise I think I'd rather move anything runtime-specific away from generic code.

Reply:
I can move it to the docker code. I was thinking that for other container runtimes the label would still be serialized json with a known structure but this does combine  the step of parsing the serialized json *and* grabbing the label, so it is still a bit docker specific.

Reply:
CRI uses gRPC instead of JSON (even though there are JSON-formatted fields in the spec).

I don't fully grok your patch but I feel we should have the runtime-specific code parse the data from whatever native format it comes in and then container_info.cpp should only handle runtime-agnostic data.

Reply:
Ok, I reorganized things so the json representation of the container in the PPME_CONTAINER_JSON event is completely divorced from how the container might happen to convey healthcheck/liveness probe/readiness probe information. All the parsing of the json from the docker daemon is isolated to the docker engine resolver. The CONTAINER_JSON json only has properties for healthcheck/liveness probe/readiness probe with an exe + args.